### PR TITLE
The pull request for the Enumerable Methods features contains the following changes

### PR DESCRIPTION
**Pull request by tiagorahal**
### The pull request for the Bubble_sort features contains the following changes
* ## Added features 
* Added enumerable.rb file
* Configured or install gem rubocop for checking the errors
*  Added .rubocop.yml
* Added .linters.yml
*  Added README.md file
*  Added MIT.md file

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
+module Enumerable
+  def my_each
```
Comment (smcommits) [inline enumerable.rb:2]: - [x] **#my_each should return an Enumerator if no block is given**
Your Output: ` LocalJumpError: no block given (yield)`

- [x] **enumerables #my_each when a block is given when self is an array should return the array itself after calling the given block once for each element in self**

- [x] **enumerables #my_each when a block is given when self is a range should return the range itself after calling the given block once for each element in self.**
Your Output: 
```NameError: undefined local variable or method `length' for 5..50:Range```

- [x] **enumerables #my_each when a block is given when self is a range behaves like #each calls the given block once for each element in self**
Your Output: 
```NameError: undefined local variable or method `length' for 5..50:Range```

- [x] **enumerables #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self**
```
expected: "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"
            got: ""
```

Comment (tiagorahal) [reply enumerable.rb:2]: Hello, I believe that the changes required are done.

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    length.times do |x|
+      yield(self[x])
+    end
+  end
+
+  def my_each_with_index
```
Comment (smcommits) [inline enumerable.rb:13]: - [x] **enumerables #my_each_with_index should return an enumerator if no block is given**
Your Output: 
```
LocalJumpError:
       no block given (yield)
```

- [x] **enumerables #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self**
```
expected: [1,2,3,4,5]
            got: 15
```

- [x] ** enumerables #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self**
```
NameError:
       undefined local variable or method `length' for 5..50:Range
```

- [x] **enumerables #my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self**
```
 NameError:
       undefined local variable or method `length' for 5..50:Range
```
- [x] ** enumerables #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self **
``` 
expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
             got: "Num: , idx: 0\nNum: , idx: 1\nNum: , idx: 2\nNum: , idx: 3\nNum: , idx: 4\n"
```

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    length.times do |x|
+      yield(self[x], x)
+    end
+  end
+
+  def my_select
```
Comment (smcommits) [inline enumerable.rb:24]: - [x] **enumerables #my_select returns an array containing all elements of enum for which the given block returns a true value**
```
expected: [1, 2, 1, 1, 2, 1, 1, 0, 3, 2, 0, 3, 2, 2, 3, 1, 2, 3, 2, 1, 2, 2, 2, 1, 3, 1, 2, 0, 1, 2, 2, 3, 2, 0, 1, 2, 2, 3, 0, 2, 1, 1, 3, 2]
             got: 100

```
- [x] **enumerables #my_select returns an array containing all elements of enum for which the given block returns a true value::range**
Your Output: 
```
NameError:
        undefined local variable or method `length' for 5..50:Range
```

- [x] **enumerables #my_select returns an enumerator if no block is given**
Your output: 
```
LocalJumpError:
        no block given (yield)
```

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    length.times do |x|
+      puts self[x] if yield(self[x])
+    end
+  end
+
+  def my_all?
```
Comment (smcommits) [inline enumerable.rb]: - [x]  **enumerables #my_all? returns false if the block returns false or nil::range**
Your output: 
```
NameError:
        undefined local variable or method `length' for 5..50:Range
```
- [x] enumerables #my_all? when no block or argument is given returns true when none of the collection members are false or nil
Your Output:
```
LocalJumpError:
        no block given (yield)
```

- [x] **enumerables #my_all? when no block or argument is given returns false when one of the collection members are false or nil**
Your output: 
```
LocalJumpError:
        no block given (yield)
```

- [x]  enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
Your Output: 
```
ArgumentError:
        wrong number of arguments (given 1, expected 0)
```
- [x] enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
Your Output: 
```
ArgumentError:
        wrong number of arguments (given 1, expected 0)
```
- [x] enumerables #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
Your Ouput: 
```
ArgumentError:
        wrong number of arguments (given 1, expected 0)
```

- [x] enumerables #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
Your Ouput: 
```
ArgumentError:
        wrong number of arguments (given 1, expected 0)
```

File: enumerable.rb (hunk 2 above)
Comment (tiagorahal) [reply enumerable.rb:13]: Hello, I believe that the changes required are done.

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    length.times do |x|
+      return false if yield(self[x]) != true
+    end
+    true
+  end
+
+  def my_any?
```
Comment (smcommits) [inline enumerable.rb]: - [x] **enumerables #my_any? returns false if the block ever returns a value false or nil::range**
Your Ouput: 
```
NameError:
        undefined local variable or method `length' for 5..50:Range
```


- [x] **enumerables #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil**
Your Ouput: 
```
LocalJumpError:
        no block given (yield)
```



- [x] **enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric**
Your Ouput: 
```
 ArgumentError:
        wrong number of arguments (given 1, expected 0)
```

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    length.times do |x|
+      return true if yield(self[x])
+    end
+    false
+  end
+
+  def my_none?
```
Comment (smcommits) [inline enumerable.rb]: - [x] **enumerables #my_none? returns false if the block ever returns true for all elements::range**
Your Ouput: 
```
NameError:
        undefined local variable or method `length' for 5..50:Range
```


- [x] **enumerables #my_none? when no block or argument is given returns true only if none of the collection members is true**
Your Ouput: 
```
LocalJumpError:
        no block given (yield)
```
- [x] **enumerables #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class**
Your Ouput: 
```
ArgumentError:
        wrong number of arguments (given 1, expected 0)
```

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    length.times do |x|
+      return false if yield(self[x])
+    end
+    true
+  end
+
+  def my_count
```
Comment (smcommits) [inline enumerable.rb]: - [x] **enumerables #my_count returns the number of items in enum through enumeration**
Your Output: 
```
 LocalJumpError:
        no block given (yield)

```

- [x] **enumerables #my_count returns the number of items in enum through enumeration::range**
Your Output: 
```
NameError:
        undefined local variable or method `length' for 5..50:Range
```

- [x] **enumerables #my_count counts the number of items in enum that are equal to item if an argument is given**
```
 ArgumentError:
        wrong number of arguments (given 1, expected 0)
```

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    counter = 0
+    length.times do |x|
+      counter += 1 if yield(self[x])
+    end
+    counter
+  end
+
+  def my_map(proc = nil)
```
Comment (smcommits) [inline enumerable.rb:89]: - [x] **enumerables #my_map returns a new array with the results of running a given block once for every element in enum.**

```
expected: [true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true...rue, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true]
             got: nil
```

- [x] **enumerables #my_map returns a new array with the results of running a given block once for every element in enum::range**
Your Output:
```
NameError:
        undefined local variable or method `length' for 5..50:Range
```

- [x] ** enumerables #my_map returns an Enumerator if no block is given**
Your Output:
```
 LocalJumpError:
        no block given (yield)
```


- [x] **#my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.**

```
expected: [false, false, false, false, false, false, false, false, false, false, false, false, false, false, fa...e, false, false, false, false, false, false, false, false, false, false, false, false, false, false]
             got: nil
```

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+      end
+    else
+      length.times do |x|
+        new_arr[x] = yield(self[x])
+      end
+    end
+    puts new_arr
+  end
+
+  def my_inject(arr = 0)
```
Comment (smcommits) [inline enumerable.rb]: - [x] **enumerables #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range**
Your Output:
```
  NameError:
        undefined local variable or method `length' for 5..50:Range
```

- [x] **enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method**
Your output:
```
LocalJumpError:
        no block given (yield)
```

- [x] **enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method**
```
ArgumentError:
        wrong number of arguments (given 2, expected 0..1)

```

**Review by smcommits — CHANGES_REQUESTED** [review]
# Status: Changes Required
## Hello, team! You have done a great job. Your methods are passing almost all the test cases. But, we still have some missing behaviours that can be found in the original Enumerable methods. 

## Required Changes
- [x] **Your methods are missing some behaviour that are present in the original Enumerable methods. Please refer to following the inline comments to understand them.**

Please use [this documentation](https://ruby-doc.org/core-3.0.0/Enumerable.html) to understand all the characterstics of enumberable methods. Just in case I miss something. 

Variables used for testing.
```
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

array = Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } 
block = proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } 
words = %w[dog door rod blade] 
range = Range.new(5, 50) 
hash = { a: 1, b: 2, c: 3, d: 4, e: 5 } 
numbers = [1, 2i, 3.14] 
array_clone = array.clone 
```

File: enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    self
+  end
+
+  def my_each_with_index
+    return enum_for(:my_each_with_index) unless block_given?
+
+    length.times do |x|
+      yield(self[x], x)
+    end
+  end
```
Comment (bini-i) [inline enumerable.rb]: ### my_each_with_index

- [x] when a block is given when self is an array returns the array itself after calling the given block once for each element in self
`array.my_each_with_index(&block)`
#### expected
`[3, 0, 0, 5, 1, 7, 8, 7, 4, 7, 6, 1, 0, 1, 1, 4, 6, 2, 7, 6, 1, 8, 8, 0, 3, 8, 6, 4, 6, 6, 0, 8, 1, 8..., 1, 7, 2, 7, 8, 0, 1, 8, 2, 0, 6, 3, 7, 3, 8, 5, 6, 0, 2, 4, 5, 2, 4, 1, 8, 8, 6, 6, 6, 7, 3, 8, 6]`
#### got
`100`

#### hint
- make sure you return the array at the end

- [x] when self is range raises the following error `undefined local variable or method 'length'`

#### hint
- As length is not a method of range, look for other methods of finding the length of a range or casting range into an array.

File: enumerable.rb [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+  def my_select
+    return enum_for(:my_select) unless block_given?
+
+    arr = []
+    to_a.length.times do |x|
+      arr.push(self[x]) if yield(self[x])
+    end
+    arr
+  end
```
Comment (bini-i) [inline enumerable.rb]: ### my_select

- [x] when self is range raises the following error `undefined method '[]' for range`
#### hint
- make sure you don't index a range using `[]`, as it is not available for range.

File: enumerable.rb [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+      to_a.my_each { |n| result = result.nil? ? n : yield(result, n) }
+    end
+    result
+  end
+end
+
+def multiply_els
+  my_inject(1) do |k, x|
+    k * x
+  end
```
Comment (bini-i) [inline enumerable.rb]: ### multiply_els

- [x] make sure it accepts an array as an argument, currently, it expects 0 arguments

**Review by bini-i — CHANGES_REQUESTED** [review]
# Status: Changes Required :yellow_circle:    
### Hello  team  :wave: 

Thanks for working on the requirements suggested by the previous TSE, but you have remaining tests that are not passing. Please refer to the inline comments below :point_down: .

### Variables used for testing

```
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

array = Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } 
block = proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } 
words = %w[dog door rod blade] 
range = Range.new(5, 50) 
hash = { a: 1, b: 2, c: 3, d: 4, e: 5 } 
numbers = [1, 2i, 3.14] 
array_clone = array.clone
```

#### If you have any question kindly reach me on slack @Binyam Hailemeskel

### HAPPY CODING :computer: :blush: 

> #### Reviewed by Binyam Hailemeskel (TSE)
> GitHub: [@bini-i](https://github.com/bini-i)
> Twitter: [@binyamshewa](https://twitter.com/binyamshewa)
> LinkedIn: [LinkedIn](https://www.linkedin.com/in/binyam-hailemeskel-728048151/)

**Review by smcommits — APPROVED** [review]
# Status: Approved :+1: 
# Hello, team! You guys have done an excellent job. You did great work addressing all the issues. This PR is now approved. You can go ahead and merge the branch. Congratulations and good luck with your next project. 

# 
Any further questions? Please reach out to me on Slack.

**Pull request merged**